pkg/storage/stores/shipper/compactor/deletion/gennumber_loader.go

```diff
@@ -0,0 +1,173 @@
+package deletion
+
+import (
+	"context"
+	"fmt"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus/promauto"
+
+	"github.com/go-kit/log/level"
+
+	"github.com/grafana/loki/pkg/util/log"
+
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+const reloadDuration = 5 * time.Minute
+
+type GenNumberLoader struct {
+	numberGetter genNumberGetter
+	numbers      map[string]string
+	quit         chan struct{}
+	lock         sync.RWMutex
+	metrics      *genLoaderMetrics
+}
+
+type genNumberGetter interface {
+	GetCacheGenerationNumber(ctx context.Context, userID string) (string, error)
+}
+
+type genLoaderMetrics struct {
+	cacheGenLoadFailures prometheus.Counter
+}
+
+func NewGenNumberLoader(g genNumberGetter, registerer prometheus.Registerer) *GenNumberLoader {
+	if g == nil {
+		g = &noopNumberGetter{}
+	}
+
+	l := &GenNumberLoader{
+		numberGetter: g,
+		numbers:      make(map[string]string),
+		metrics:      newGenLoaderMetrics(registerer),
+	}
+	go l.loop()
+
+	return l
+}
+
+func newGenLoaderMetrics(r prometheus.Registerer) *genLoaderMetrics {
+	return &genLoaderMetrics{
+		cacheGenLoadFailures: promauto.With(r).NewCounter(prometheus.CounterOpts{
+			Namespace: "loki",
+			Name:      "delete_cache_gen_load_failures_total",
+			Help:      "Total number of failures while loading cache generation number using gen number loader",
+		}),
+	}
+}
+
+func (l *GenNumberLoader) loop() {
+	timer := time.NewTicker(reloadDuration)
+	for {
+		select {
+		case <-timer.C:
+			err := l.reload()
+			if err != nil {
+				level.Error(log.Logger).Log("msg", "error reloading tombstones", "err", err)
+			}
+		case <-l.quit:
+			return
+		}
+	}
+}
+
+func (l *GenNumberLoader) reload() error {
+	updatedGenNumbers, err := l.getUpdatedGenNumbers()
+	if err != nil {
+		return err
+	}
+
+	l.lock.Lock()
+	defer l.lock.Unlock()
+	for userID, genNumber := range updatedGenNumbers {
+		l.numbers[userID] = genNumber
+	}
+
+	return nil
+}
+
+func (l *GenNumberLoader) getUpdatedGenNumbers() (map[string]string, error) {
+	l.lock.RLock()
+	defer l.lock.RUnlock()
+
+	updatedGenNumbers := make(map[string]string)
+	for userID, oldGenNumber := range l.numbers {
+		genNumber, err := l.numberGetter.GetCacheGenerationNumber(context.Background(), userID)
+		if err != nil {
+			return nil, err
+		}
+
+		if oldGenNumber != genNumber {
+			updatedGenNumbers[userID] = genNumber
+		}
+	}
+
+	return updatedGenNumbers, nil
+}
+
+func (l *GenNumberLoader) GetResultsCacheGenNumber(tenantIDs []string) string {
+	return l.getCacheGenNumbersPerTenants(tenantIDs)
+}
+
+func (l *GenNumberLoader) GetStoreCacheGenNumber(tenantIDs []string) string {
+	return l.getCacheGenNumbersPerTenants(tenantIDs)
+}
```

Comment:
I'm still a little confused on this. Why not just invalidate all the caches whenever a delete happens? Is the idea to preserve the storage cache as long as possible?

Reply:
I too am confused. If we invalidate both, won't the results cache get updated as a result of caching the results from the query that hit the actual store on account of the store cache also being invalidated? That seems like less complexity to me, just always invalidate both.

Reply:
When a new delete request comes in, we will do query time filtering of data until the actual data from the storage gets deleted. So if we invalidate both results and store cache when a new delete request comes in, we would anyways cache the same data since we have not processed the delete requests yet, so it would be a waste to drop and cache the same data again.

Similarly, when the delete request is processed, whatever results we have cached would be built after filtering out deleted data since we would have cleared the results cache already when the delete request was received. So, if we invalidate the results cache again after deleting data from the store, we would be wasting resources to drop and cache the same data again, this time without query time filtering but with the actual data deleted from storage.

Reply:
This makes sense. I'm going to implement it in a follow on PR